orangecontrib/text/preprocess.py

```diff
@@ -1,94 +1,298 @@
-        if isinstance(data, list):
-            vec = self.cv.fit(data)
-            features = vec.get_feature_names()
-            docs = [[] for _ in range(len(data))]
-            for (line, column), count in vec.transform(data).todok().items():
-                docs[line].extend([features[column]] * count)
-            return docs
+            return self.tokenizer(data)
         else:
-            raise ValueError("Type '{}' not supported.".format(type(data)))
+            raise ValueError('Type {} not supported.'.format(type(data)))
+
+    def _transform(self, data):
+        """Performs whatever transformation was specified at instance creation.
+
+        Args:
+            data (str, list): The input holding the text to be transformed.
+                Can be provided in form of a string or list of string tokens.
+
+        Returns:
+            str, list: Depending on the input, returns either a list of tokens
+                or a single token.
+        """
+        return self.transformation(data)
+
+    @staticmethod
+    def _remove_stop_words(data, stop_words):
+        """Removes the stop words from the corpus.
+
+        Removes the stop words specified in the stop_words parameter.
+
+        Args:
+            data (list): The input document in the form of a list.
+            stop_words (set): The set of stop word we wish to use.
+
+        Returns:
+            list: Pre-processed document in the form of a list, holding the
+                tokens.
+        """
+        assert isinstance(stop_words, set)
+        return [word for word in data if word not in stop_words]
+
+    def _get_corpus_specific_stopwords(self, corpus_documents):
+        """Calculates the corpus specific stop words.
+
+        Constructs a stop word list, using the minimum and maximum document
+        frequency parameters.
+
+        Args:
+            corpus_documents (list): The input corpus structured as a list of
+                lists.
+
+        Returns:
+            list: Document frequency stop words in a list.
+        """
+        # 1.) First, we need to tokenize the documents in the corpus.
+        corpus_documents = [self._tokenize(doc) for doc in corpus_documents]
+
+        document_frequencies = {}
+        # 2.) Calculate the document frequency for individual tokens.
+        for document in corpus_documents:
+            for token in set(document):
+                document_frequencies[token] = document_frequencies.get(
+                        token, 0) + 1
+
+        # 3.) SW based on minimum document frequency.
+        sw_min = set()
+        if self.min_df is not None:
+            min_df = self.min_df  # If int.
+            if isinstance(self.min_df, float):  # If float.
+                min_df = round(len(corpus_documents) * self.min_df)
+            sw_min = {token for token, freq in document_frequencies.items()
+                      if freq <= min_df}
+
+        # 4.) SW based on maximum document frequency.
```

Comment:
Can you move this into section 3? We can check: `{token for token, freq in document_frequencies.items() if freq < min_df or freq > max_df}`?